crates/subspace-farmer/src/dsn.rs

```diff
@@ -0,0 +1,124 @@
+use crate::PiecesToPlot;
+use futures::{Stream, StreamExt};
+use std::ops::{ControlFlow, Range};
+use std::sync::{Arc, Mutex};
+use subspace_core_primitives::{FlatPieces, PieceIndex, PieceIndexHash, PublicKey, Sha256Hash};
+use subspace_solving::U256;
+
+#[cfg(test)]
+mod tests;
+
+pub type PieceIndexHashNumber = U256;
+
+/// Options for syncing
+pub struct SyncOptions {
+    /// The size of request in pieces which we should request on average
+    pub pieces_per_request: u64,
+    /// The size of range which we request initially
+    pub initial_range_size: PieceIndexHashNumber,
+    /// The size of range at which we should stop syncing
+    pub max_range_size: PieceIndexHashNumber,
+    /// Address of our plot
+    pub address: PublicKey,
+}
+
+#[async_trait::async_trait]
+pub trait DSNSync {
+    type Stream: Stream<Item = PiecesToPlot>;
+
+    /// Get pieces from the network which fall within the range
+    async fn get_pieces(&mut self, range: Range<PieceIndexHash>) -> Self::Stream;
+}
+
+/// Marker which disables sync
+#[derive(Clone, Copy, Debug)]
+pub struct NoSync;
+
+#[async_trait::async_trait]
+impl DSNSync for NoSync {
+    type Stream = futures::stream::Empty<PiecesToPlot>;
+
+    async fn get_pieces(&mut self, _range: Range<PieceIndexHash>) -> Self::Stream {
+        futures::stream::empty()
+    }
+}
+
+// TODO: Can we know in advance amount of all pieces in the network? That would simplify the whole
+// algorithm as it would simply request all pieces within expected range of pieces for which we can
+// get rewards from the consensus.
+/// Syncs pieces from the provided DSN.
+///
+/// It would query pieces going upwards and downwards. It will adjust the requested piece range in
+/// order to ask the same amount of data over the network.
```

Comment:
Are you sure this is needed? What we have discussed with @shamil-gadelshin was something like this:
* make a request for pieces in range `A..=B`
* get response with `N` first pieces in the range + `cursor`
* request next `N` pieces using the same range and cursor until download the whole range

Reply:
Makes sense. I'll rewrite it in this manner

Reply:
I think we can hide a cursor inside our internal implementation of dsn and just provide a stream to the farmer for syncing